src/Symfony/Component/Translation/Extractor/Visitor/ConstraintVisitor.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Extractor\Visitor;
+
+use PhpParser\Node;
+use PhpParser\NodeVisitor;
+
+/**
+ * @author Mathieu Santostefano <msantostefano@protonmail.com>
+ *
+ * Code mostly comes from https://github.com/php-translation/extractor/blob/master/src/Visitor/Php/Symfony/Constraint.php
+ */
+final class ConstraintVisitor extends AbstractVisitor implements NodeVisitor
+{
+    public function __construct(
+        private readonly array $constraintClassNames = []
+    ) {
+    }
+
+    public function beforeTraverse(array $nodes): ?Node
+    {
+        return null;
+    }
+
+    public function enterNode(Node $node): ?Node
+    {
+        if (!$node instanceof Node\Expr\New_ && !$node instanceof Node\Attribute) {
+            return null;
+        }
+
+        $className = $node instanceof Node\Attribute ? $node->name : $node->class;
+        if (!$className instanceof Node\Name) {
+            return null;
+        }
+
+        $parts = $className->parts;
+        $isConstraintClass = false;
+
+        foreach ($parts as $part) {
+            if (\in_array($part, $this->constraintClassNames, true)) {
+                $isConstraintClass = true;
+
+                break;
+            }
+        }
+
+        if (!$isConstraintClass) {
+            return null;
+        }
+
+        if ($node instanceof Node\Expr\New_) {
+            $this->extractFromConstraintClassInstantiation($node);
+        } else {
+            $this->extractFromConstraintAttribute($node);
+        }
+
+        return null;
+    }
+
+    public function leaveNode(Node $node): ?Node
+    {
+        return null;
+    }
+
+    public function afterTraverse(array $nodes): ?Node
+    {
+        return null;
+    }
+
+    private function extractFromConstraintClassInstantiation(Node\Expr\New_ $node): void
```

Comment:
This currently only supports extracting from the option array. We also need to support extracting from named constructor arguments IMO.

Reply:
Indeed, I take care of it.

Reply:
Good catch, I just noticed that I miss some major cases here (i.e if options are passed traditionally, not in an array).